lk2nd/device/leds.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright (c) 2025 Yang Xiwen <forbidden405@outlook.com> */
+
+#include <debug.h>
+
+#include <libfdt.h>
+
+#include <lk2nd/hw/gpio.h>
+
+#include "device.h"
+
+#define LK2ND_MAX_LEDS 16
+
+enum led_state {
+	LED_STATE_UNKNOWN = -1,
+	LED_STATE_ON,
+	LED_STATE_OFF,
+	LED_STATE_KEEP,
+};
+
+static int lk2nd_leds_init(const void *dtb, int node)
+{
+	int i = 0, subnode, len;
+	struct gpiol_desc desc;
+	const char *state;
+	enum led_state estate = LED_STATE_UNKNOWN;
+
+	dprintf(SPEW, " Initializing LEDs\n");
+	dprintf(SPEW, " | label | code  | gpio        |\n");
+
+	fdt_for_each_subnode(subnode, dtb, node) {
+		if (i == LK2ND_MAX_LEDS) {
+			dprintf(CRITICAL, "leds: Too many leds defined, ignoring rest\n");
+			break;
+		}
+
+		state = fdt_getprop(dtb, subnode, "default-state", &len);
+		if (len < 0 || state[len - 1] != '\0') {
+			dprintf(CRITICAL, "leds: Failed to get property \"default-state\". Broken dtb?\n");
+			continue;
+		}
+
+		if (!strcmp(state, "on"))
+			estate = LED_STATE_ON;
+		else if (!strcmp(state, "off"))
+			estate = LED_STATE_OFF;
+		else if (!strcmp(state, "keep"))
+			estate = LED_STATE_KEEP;
+
+		switch (estate) {
+		case LED_STATE_ON:
```

Comment:
TBH considering this state value isn't used anywhere else, I'd just fold this switch and if/elif above and set the gpio immediately in "on/off"

Reply:
ok. I've also removed the enum.